Limit in GraphCollections misleading

The documentation of [GraphCollectionOperators#limit](https://github.com/dbs-leipzig/gradoop/blob/master/gradoop-flink/src/main/java/org/gradoop/flink/model/api/epgm/GraphCollectionOperators.java#L90-L97) says it returns a restricted number of graph collections. The implementation of InAllGraphsBroadcast[](https://github.com/dbs-leipzig/gradoop/blob/master/gradoop-flink/src/main/java/org/gradoop/flink/model/impl/functions/graphcontainment/InAllGraphsBroadcast.java#L33) filters too many vertices and edges: It only allows elements that are contained in ALL filtered graph heads, not just any of them.

Here a test that fails with the message ```"java.lang.AssertionError: expected:<2> but was:<1>"``` (can be inserted into [LimitTest](https://github.com/dbs-leipzig/gradoop/blob/master/gradoop-flink/src/test/java/org/gradoop/flink/model/impl/operators/limit/LimitTest.java)):

```java
  @Test
  public void testDistinctGraphs() throws Exception {
    FlinkAsciiGraphLoader loader = getLoaderFromString("(e1) (e2) g1 [(e1)] g2 [(e1) (e2)]");

    int limit = 2;

    GraphCollection inputCollection = loader
      .getGraphCollectionByVariables("g1", "g2");

    GraphCollection outputCollection =
      inputCollection.limit(limit);

    assertEquals(limit, outputCollection.getGraphHeads().count());
    assertEquals(2, outputCollection.getVertices().count());
  }

--- a/gradoop-flink/src/main/java/org/gradoop/flink/model/impl/operators/limit/Limit.java
+++ b/gradoop-flink/src/main/java/org/gradoop/flink/model/impl/operators/limit/Limit.java
@@ -24,7 +24,7 @@ import org.gradoop.flink.model.api.epgm.GraphCollection;
 import org.gradoop.flink.model.api.operators.UnaryCollectionToCollectionOperator;
 import org.gradoop.flink.model.impl.functions.epgm.Id;
 import org.gradoop.flink.model.impl.functions.graphcontainment.GraphsContainmentFilterBroadcast;
-import org.gradoop.flink.model.impl.functions.graphcontainment.InAllGraphsBroadcast;
+import org.gradoop.flink.model.impl.functions.graphcontainment.InAnyGraphBroadcast;
 
 /**
  * Returns the first n (arbitrary) logical graphs from a collection.
@@ -53,14 +53,14 @@ public class Limit implements UnaryCollectionToCollectionOperator {
 
     DataSet<GraphHead> graphHeads = collection.getGraphHeads().first(limit);
 
-    DataSet<GradoopId> firstIds = graphHeads.map(new Id<GraphHead>());
+    DataSet<GradoopId> firstIds = graphHeads.map(new Id<>());
 
     DataSet<Vertex> filteredVertices = collection.getVertices()
-      .filter(new InAllGraphsBroadcast<Vertex>())
+      .filter(new InAnyGraphBroadcast<>())
       .withBroadcastSet(firstIds, GraphsContainmentFilterBroadcast.GRAPH_IDS);
 
     DataSet<Edge> filteredEdges = collection.getEdges()
-      .filter(new InAllGraphsBroadcast<Edge>())
+      .filter(new InAnyGraphBroadcast<>())
       .withBroadcastSet(firstIds, GraphsContainmentFilterBroadcast.GRAPH_IDS);
 
     return collection.getConfig().getGraphCollectionFactory()

--- a/gradoop-flink/src/test/java/org/gradoop/flink/model/impl/operators/limit/LimitTest.java
+++ b/gradoop-flink/src/test/java/org/gradoop/flink/model/impl/operators/limit/LimitTest.java
@@ -80,4 +80,18 @@ public class LimitTest extends GradoopFlinkTestBase {
 
     assertEquals(expectedCount, outputCollection.getGraphHeads().count());
   }
+
+  @Test
+  public void testDistinctGraphs() throws Exception {
+    FlinkAsciiGraphLoader loader = getLoaderFromString("(v1) (v2) g1[(v1)] g2[(v1) (v2)]");
+
+    int limit = 2;
+
+    GraphCollection inputCollection = loader.getGraphCollectionByVariables("g1", "g2");
+
+    GraphCollection outputCollection = inputCollection.limit(limit);
+
+    assertEquals(limit, outputCollection.getGraphHeads().count());
+    assertEquals(2, outputCollection.getVertices().count());
+  }
 }

--- a/gradoop-flink/src/test/java/org/gradoop/flink/model/impl/operators/limit/LimitTest.java
+++ b/gradoop-flink/src/test/java/org/gradoop/flink/model/impl/operators/limit/LimitTest.java
@@ -19,11 +19,23 @@ import org.apache.flink.api.common.InvalidProgramException;
 import org.gradoop.flink.model.GradoopFlinkTestBase;
 import org.gradoop.flink.model.api.epgm.GraphCollection;
 import org.gradoop.flink.util.FlinkAsciiGraphLoader;
+import org.jamon.annotations.Template;
 import org.junit.Test;
+import org.s1ck.gdl.model.Graph;
 
 import static org.junit.Assert.assertEquals;
 
 public class LimitTest extends GradoopFlinkTestBase {
+
+  @Test
+  public void testSameCollection() throws Exception {
+    FlinkAsciiGraphLoader loader = getSocialNetworkLoader();
+
+    int limit = 2;
+
+    GraphCollection input = loader.getGraphCollectionByVariables("g0", "g1");
+    collectAndAssertTrue(input.equalsByGraphElementData(input.limit(limit)));
+  }
 
   @Test
   public void testInBound() throws Exception {
@@ -31,12 +43,9 @@ public class LimitTest extends GradoopFlinkTestBase {
 
     int limit = 2;
 
-    GraphCollection inputCollection = loader
-      .getGraphCollectionByVariables("g0", "g1", "g2", "g3");
+    GraphCollection inputCollection = loader.getGraphCollectionByVariables("g0", "g1", "g2", "g3");
 
-    GraphCollection outputCollection =
-      inputCollection.limit(limit);
-
+    GraphCollection outputCollection = inputCollection.limit(limit);
     assertEquals(limit, outputCollection.getGraphHeads().count());
   }
 
@@ -44,15 +53,13 @@ public class LimitTest extends GradoopFlinkTestBase {
   public void testOutOfBound() throws Exception {
     FlinkAsciiGraphLoader loader = getSocialNetworkLoader();
 
-    GraphCollection inputCollection = loader
-      .getGraphCollectionByVariables("g0", "g1");
+    GraphCollection inputCollection = loader.getGraphCollectionByVariables("g0", "g1");
 
     int limit = 4;
-    int expectedLimit = 2;
+    int expectedCount = 2;
 
     GraphCollection outputCollection = inputCollection.limit(limit);
-
-    assertEquals(expectedLimit, outputCollection.getGraphHeads().count());
+    assertEquals(expectedCount, outputCollection.getGraphHeads().count());
   }
 
   @Test
@@ -64,7 +71,6 @@ public class LimitTest extends GradoopFlinkTestBase {
     int expectedCount = 0;
 
     GraphCollection outputCollection = inputCollection.limit(limit);
-
     assertEquals(expectedCount, outputCollection.getGraphHeads().count());
   }
 
@@ -77,7 +83,6 @@ public class LimitTest extends GradoopFlinkTestBase {
     int expectedCount = 0;
 
     GraphCollection outputCollection = inputCollection.limit(limit);
-
     assertEquals(expectedCount, outputCollection.getGraphHeads().count());
   }
 
@@ -88,10 +93,6 @@ public class LimitTest extends GradoopFlinkTestBase {
     int limit = 2;
 
     GraphCollection inputCollection = loader.getGraphCollectionByVariables("g1", "g2");
-
-    GraphCollection outputCollection = inputCollection.limit(limit);
-
-    assertEquals(limit, outputCollection.getGraphHeads().count());
-    assertEquals(2, outputCollection.getVertices().count());
+    collectAndAssertTrue(inputCollection.equalsByGraphElementData(inputCollection.limit(limit)));
   }
 }

